test/src/tests-session-manager.ts

```diff
@@ -0,0 +1,622 @@
+            const now = new Date();
+            // The default timeout limit is 30 minutes.
+            const twentyMinutesAgo = new Date();
+            twentyMinutesAgo.setMinutes(now.getMinutes() - 20);
+
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+            const timerSpy = sinon.spy(
+                mpInstance._SessionManager,
+                'setSessionTimer'
+            );
+
+            // Session Manager relies on persistence to determine last time seen (LES)
+            // Also requires sid to verify session exists
+            sinon.stub(mpInstance._Persistence, 'getPersistence').returns({
+                gs: {
+                    les: twentyMinutesAgo,
+                    sid: 'fake-session-id',
+                },
+            });
+
+            mpInstance._SessionManager.endSession();
+
+            expect(mpInstance._Store.sessionId).to.equal('fake-session-id');
+
+            // When session is not timed out, setSessionTimer is called to keep track
+            // of current session timeout
+            expect(timerSpy.getCalls().length).to.equal(1);
+        });
+
+        it('should end session if session times out', () => {
+            const generateUniqueIdSpy = sinon.stub(
+                mParticle.getInstance()._Helpers,
+                'generateUniqueId'
+            );
+            generateUniqueIdSpy.returns('test-unique-id');
+
+            const now = new Date();
+            const hourAgo = new Date();
+            hourAgo.setMinutes(now.getMinutes() - 60);
+
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+
+            expect(mpInstance._Store.sessionId).to.equal('TEST-UNIQUE-ID');
+            // Init will set dateLastEventSent to now, but endSession relies on the persistence layer
+            expect(mpInstance._Store.dateLastEventSent).to.eql(now);
+            expect(mpInstance._Store.sessionAttributes).to.eql({});
+
+            const persistenceSpy = sinon.spy(mpInstance._Persistence, 'update');
+
+            // Session Manager relies on persistence to determine last time seen (LES)
+            // Also requires sid to verify session exists
+            sinon.stub(mpInstance._Persistence, 'getPersistence').returns({
+                gs: {
+                    les: hourAgo,
+                    sid: 'fake-session-id',
+                },
+            });
+
+            mpInstance._SessionManager.endSession();
+
+            expect(mpInstance._Store.sessionId).to.equal(null);
+            expect(mpInstance._Store.dateLastEventSent).to.equal(null);
+            expect(mpInstance._Store.sessionAttributes).to.eql({});
+
+            // Persistence isn't necessary for this feature, but we should test
+            // to see that it is called in case this ever needs to be refactored
+            expect(persistenceSpy.called).to.equal(true);
+        });
```

Comment:
I think I understand your reasoning, but my intention is to have these tests be as true "unit tests" as we possibly can with the SDK. I think testing the behavior of ticking the clock forward is very much an integration test, and possibly should be moved to a separate section of this test suite. However, I see value in highlighting expected behavior of calling `endSession` in all it's permutations.

I think for now I will separate Unit and Integration tests, and add the clock ticking test as an Integration test, since it does not directly rely on calling `endSession`. We can then group similar tests for clarity in the future.